### ISSUE TITLE
System Time Change: ZK connection loss - Leadership Deadlock & Socket Leaks

### ISSUE THREAD

**Issue by Brycelol** [mesosphere/marathon#4624]
**OS:** Red Hat Enterprise Linux Server release 7.2 (Maipo)
**Kernel:** 3.10.0-327.el7.x86_64
**Marathon Version:**  1.3.0

In _some_ cases, when modifying the system time slightly, the Zookeeper session exceeds the specified (10s) timeout and causes the leadership election service to deadlock rather than commit suicide or abdicate leadership. When it enters this deadlocked state, any new connections (typical health check pings) which lead Jetty to spawn new handling threads also get deadlocked attempting to use the election service causing socket leaks.

**Causes of the Deadlock**

ElectionServiceBase.stopLeadership() is called by the Curator-ConnectionStateManager-0 thread.  It grabs a lock on itself (a CuratorElectionService instance in this case), then calls candidate.stopLeadership(), which is MarathonSchedulerService.stopLeadership().  That method needs to grab a lock on the MarathonSchedulerService instance.

Meanwhile, thread ForkJoinPool-2-worker-11 has called MarathonSchedulerService.startLeadership() which starts by grabbing a lock on the MarathonSchedulerService instance, which the other thread is waiting for.  Then it loops through drivers calling abdicateLeadership() on them.  In this case, this is the ElectionServiceBase.abdicateLeadership() method that requires a lock on the CuratorElectionService that the other thread already has.

Voila, deadlock.  Then all the servlet requests get blocked waiting for a lock on the CuratorElectionService instance that will never be freed because of the deadlock.  Voila, socket leak.

**A State Dump shows the deadlock from a sample JVM**

```
1LKDEADLOCK    Deadlock detected !!!  
NULL           ---------------------  
NULL             
2LKDEADLOCKTHR  Thread "Curator-ConnectionStateManager-0" (0x0000000001F87B00)
3LKDEADLOCKWTR    is waiting for:
4LKDEADLOCKMON      sys_mon_t:0x00007EFD044E7CB8 infl_mon_t: 0x00007EFD044E7D30:
4LKDEADLOCKOBJ      mesosphere/marathon/MarathonSchedulerService$$EnhancerByGuice$$d9cdc13b@0x00000000E0C40008
3LKDEADLOCKOWN    which is owned by:
2LKDEADLOCKTHR  Thread "ForkJoinPool-2-worker-11" (0x0000000001D3CC00)
3LKDEADLOCKWTR    which is waiting for:
4LKDEADLOCKMON      sys_mon_t:0x00007EFD044E76D0 infl_mon_t: 0x00007EFD044E7748:
4LKDEADLOCKOBJ      mesosphere/marathon/core/election/impl/CuratorElectionService@0x00000000E0BC5E20
3LKDEADLOCKOWN    which is owned by:
2LKDEADLOCKTHR  Thread "Curator-ConnectionStateManager-0" (0x0000000001F87B00)
```

**JMX Stack Trace for deadlocked Curator-ConnectionStateManager-0**

```
    Name: Curator-ConnectionStateManager-0
    State: BLOCKED on mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$c453a1e2@f09af3a owned by: ForkJoinPool-2-worker-15
    Total blocked: 4  Total waited: 73

    Stack trace: 
    mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$c453a1e2.CGLIB$stopLeadership$26(<generated>)
    mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$c453a1e2$$FastClassByGuice$$2639ebd3.invoke(<generated>)
    com.google.inject.internal.cglib.proxy.$MethodProxy.invokeSuper(MethodProxy.java:228)
    com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:72)
    mesosphere.marathon.DebugModule$MetricsBehavior$$anonfun$invoke$1.apply(DebugConf.scala:75)
    mesosphere.marathon.metrics.Metrics.timed(Metrics.scala:27)
    mesosphere.marathon.DebugModule$MetricsBehavior.invoke(DebugConf.scala:74)
    com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:72)
    com.google.inject.internal.InterceptorStackCallback.intercept(InterceptorStackCallback.java:52)
    mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$c453a1e2.stopLeadership(<generated>)
    mesosphere.marathon.core.election.impl.ElectionServiceBase.stopLeadership(ElectionServiceBase.scala:165)
       - locked mesosphere.marathon.core.election.impl.CuratorElectionService@4947a204
    mesosphere.marathon.core.election.impl.CuratorElectionService$Listener$.notLeader(CuratorElectionService.scala:74)
       - locked mesosphere.marathon.core.election.impl.CuratorElectionService@4947a204
    org.apache.curator.framework.recipes.leader.LeaderLatch$8.apply(LeaderLatch.java:640)
    org.apache.curator.framework.recipes.leader.LeaderLatch$8.apply(LeaderLatch.java:636)
    org.apache.curator.framework.listen.ListenerContainer$1.run(ListenerContainer.java:93)
    com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:310)
    org.apache.curator.framework.listen.ListenerContainer.forEach(ListenerContainer.java:85)
    org.apache.curator.framework.recipes.leader.LeaderLatch.setLeadership(LeaderLatch.java:635)
       - locked org.apache.curator.framework.recipes.leader.LeaderLatch@9f76e43f
    org.apache.curator.framework.recipes.leader.LeaderLatch.handleStateChange(LeaderLatch.java:623)
    org.apache.curator.framework.recipes.leader.LeaderLatch.access$000(LeaderLatch.java:64)
    org.apache.curator.framework.recipes.leader.LeaderLatch$1.stateChanged(LeaderLatch.java:82)
    org.apache.curator.framework.state.ConnectionStateManager$2.apply(ConnectionStateManager.java:259)
    org.apache.curator.framework.state.ConnectionStateManager$2.apply(ConnectionStateManager.java:255)
    org.apache.curator.framework.listen.ListenerContainer$1.run(ListenerContainer.java:93)
    com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:310)
    org.apache.curator.framework.listen.ListenerContainer.forEach(ListenerContainer.java:85)
    org.apache.curator.framework.state.ConnectionStateManager.processEvents(ConnectionStateManager.java:253)
    org.apache.curator.framework.state.ConnectionStateManager.access$000(ConnectionStateManager.java:43)
    org.apache.curator.framework.state.ConnectionStateManager$1.call(ConnectionStateManager.java:111)
    java.util.concurrent.FutureTask.run(FutureTask.java:277)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    java.lang.Thread.run(Thread.java:785)
```

**JMX Stack Trace for deadlocked ForkJoinPool-2-worker-15**

```
    Name: ForkJoinPool-2-worker-15
    State: BLOCKED on mesosphere.marathon.core.election.impl.CuratorElectionService@4947a204 owned by: Curator-ConnectionStateManager-0
    Total blocked: 1  Total waited: 65

    Stack trace: 
    mesosphere.marathon.MarathonSchedulerService$$anonfun$startLeadership$3$$anonfun$apply$1.apply(MarathonSchedulerService.scala:253)
    mesosphere.marathon.MarathonSchedulerService$$anonfun$startLeadership$3$$anonfun$apply$1.apply(MarathonSchedulerService.scala:251)
    scala.Option.foreach(Option.scala:257)
    mesosphere.marathon.MarathonSchedulerService$$anonfun$startLeadership$3.apply(MarathonSchedulerService.scala:251)
       - locked mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$c453a1e2@f09af3a
    mesosphere.marathon.MarathonSchedulerService$$anonfun$startLeadership$3.apply(MarathonSchedulerService.scala:238)
    scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
    scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
    scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
    scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
    scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**JVM Stack Trace of Deadlocked Jetty Servlet Thread**

```
Name: qtp584257757-51
State: BLOCKED on mesosphere.marathon.core.election.impl.CuratorElectionService@4947a204 owned by: Curator-ConnectionStateManager-0
Total blocked: 2  Total waited: 31

Stack trace: 
mesosphere.marathon.api.LeaderProxyFilter.doFilter(LeaderProxyFilter.scala:99)
com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:118)
com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:113)
org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581)
org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1158)
org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1090)
org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:240)
org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:109)
org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:437)
org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
org.eclipse.jetty.server.Server.handle(Server.java:517)
org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:308)
org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:261)
org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
java.lang.Thread.run(Thread.java:785)
```

**Comment by jasongilanfarr** [comment]
I'll take a bigger look later today, but I haven't crossported the 1.3 "exit on loss of leadership" to master yet. Wouldn't that eliminate the need for this patch?

**Comment by jasongilanfarr** [comment]
Have you tried on 1.3.x again after we added the abort? if that does fix it, maybe we can mark as a duplicate of #4639

**Comment by timcharper** [comment]
this command landed for the next release of 1.3 a7f4138a1609e410901095cce6cdcc81541206d8 and should solve the deadlock

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-1723. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
